### ISSUE TITLE
Adding fixnum-bit? intrinsic

### ISSUE THREAD

**Issue by mrjbq7** [slavapestov/factor#1367]
This doesn't seem to be a performance improvement, but I'm keeping this here so I can look at it later.

```diff
diff --git a/basis/compiler/cfg/instructions/instructions.factor b/basis/compiler/cfg/instructions/instructions.factor
index 703294e..1c44560 100644
--- a/basis/compiler/cfg/instructions/instructions.factor
+++ b/basis/compiler/cfg/instructions/instructions.factor
@@ -224,6 +224,11 @@ FOLDABLE-INSN: ##bit-count
 def: dst/int-rep
 use: src/int-rep ;
 
+FOLDABLE-INSN: ##bit-test
+def: dst/tagged-rep
+use: src1/int-rep src2/int-rep
+temp: temp/int-rep ;
+
 ! Float arithmetic
 FOLDABLE-INSN: ##add-float
 def: dst/double-rep
diff --git a/basis/compiler/cfg/intrinsics/intrinsics.factor b/basis/compiler/cfg/intrinsics/intrinsics.factor
index 7d744e7..24b7e05 100644
--- a/basis/compiler/cfg/intrinsics/intrinsics.factor
+++ b/basis/compiler/cfg/intrinsics/intrinsics.factor
@@ -138,5 +138,10 @@ ERROR: inline-intrinsics-not-supported word quot ;
         { math.bitwise.private:fixnum-bit-count [ drop [ ^^bit-count ] unary-op ] }
     } enable-intrinsics ;
 
+: enable-bit-test ( -- )
+    {
+        { math.integers.private:fixnum-bit? [ drop [ ^^bit-test ] binary-op ] }
+    } enable-intrinsics ;
+
 : emit-intrinsic ( node word -- )
     "intrinsic" word-prop call( node -- ) ;
diff --git a/basis/compiler/codegen/codegen.factor b/basis/compiler/codegen/codegen.factor
index a488a36..11175ec 100755
--- a/basis/compiler/codegen/codegen.factor
+++ b/basis/compiler/codegen/codegen.factor
@@ -181,6 +181,7 @@ CODEGEN: ##not %not
 CODEGEN: ##neg %neg
 CODEGEN: ##log2 %log2
 CODEGEN: ##bit-count %bit-count
+CODEGEN: ##bit-test %bit-test
 CODEGEN: ##copy %copy
 CODEGEN: ##tagged>integer %tagged>integer
 CODEGEN: ##add-float %add-float
diff --git a/basis/cpu/architecture/architecture.factor b/basis/cpu/architecture/architecture.factor
index 6cd42c5..7ab4957 100644
--- a/basis/cpu/architecture/architecture.factor
+++ b/basis/cpu/architecture/architecture.factor
@@ -271,6 +271,7 @@ HOOK: %not     cpu ( dst src -- )
 HOOK: %neg     cpu ( dst src -- )
 HOOK: %log2    cpu ( dst src -- )
 HOOK: %bit-count cpu ( dst src -- )
+HOOK: %bit-test cpu ( dst src1 src2 temp -- )
 
 HOOK: %copy cpu ( dst src rep -- )
 
diff --git a/basis/cpu/x86/x86.factor b/basis/cpu/x86/x86.factor
index aa9d768..9d67c93 100644
--- a/basis/cpu/x86/x86.factor
+++ b/basis/cpu/x86/x86.factor
@@ -764,6 +764,12 @@ M: x86 immediate-bitwise? ( n -- ? )
 enable-min/max
 enable-log2
 
+M:: x86 %bit-test ( dst src1 src2 temp -- )
+    src1 src2 BT
+    dst temp \ CMOVB (%boolean) ;
+
+enable-bit-test
+
 : check-sse ( -- )
     "Checking for multimedia extensions... " write flush
     sse-version
diff --git a/core/math/integers/integers.factor b/core/math/integers/integers.factor
index 98ca157..7aae3c5 100644
--- a/core/math/integers/integers.factor
+++ b/core/math/integers/integers.factor
@@ -53,7 +53,7 @@ M: fixnum bitnot fixnum-bitnot ; inline
 
 : fixnum-bit? ( x n -- ? )
     { fixnum fixnum } declare
-    dup 0 >= [ neg shift even? not ] [ 2drop f ] if ; inline
+    dup 0 >= [ neg shift even? not ] [ 2drop f ] if ;
 
 M: fixnum bit? integer>fixnum-strict fixnum-bit? ; inline
```

**Comment by mrjbq7** [comment]
This speeds up ``fixnum-bit?`` a lot:

Using:

```
: foo ( x n -- ? )
    [ 0 100,000,000 ] 2dip
    '[ _ _ bit? [ 1 + ] when ] times ;

: bar ( x n -- ? )
    { fixnum fixnum } declare
    [ 0 100,000,000 ] 2dip
    '[ _ _ bit? [ 1 + ] when ] times ;
```

Old:

```
IN: scratchpad gc [ 0b101010101 0 foo ] time
Running time: 1.163613351 seconds

IN: scratchpad gc [ 0b101010101 0 bar ] time
Running time: 0.821433838 seconds
```

New:

```
IN: scratchpad gc [ 0b101010101 0 foo ] time
Running time: 0.667124186 seconds

IN: scratchpad gc [ 0b101010101 0 bar ] time
Running time: 0.239439108 seconds
```

I'm inclined to commit it, @erg what do you think?

**Comment by mrjbq7** [comment]
Old disassemble:

```
IN: scratchpad M\ fixnum bit? disassemble
000000010ea00250: 8905aa0de0ff          mov [rip-0x1ff256], eax
000000010ea00256: 4883ec08              sub rsp, 0x8
000000010ea0025a: e851f14d00            call 0x10eedf3b0 (integer>fixnum-strict)
000000010ea0025f: 498b06                mov rax, [r14]
000000010ea00262: 4883f800              cmp rax, 0x0
000000010ea00266: 0f8c3b000000          jl 0x10ea002a7 (M\ fixnum bit? + 0x57)
000000010ea0026c: 48f7d8                neg rax
000000010ea0026f: 498906                mov [r14], rax
000000010ea00272: e8e9bd1e00            call 0x10ebec060 (fixnum-shift)
000000010ea00277: 498b06                mov rax, [r14]
000000010ea0027a: 48f7c010000000        test rax, 0x10
000000010ea00281: 48b80100000000000000  mov rax, 0x1
000000010ea0028b: 48bb4cff0e0101000000  mov rbx, 0x1010eff4c
000000010ea00295: 480f45c3              cmovnz rax, rbx
000000010ea00299: 498906                mov [r14], rax
000000010ea0029c: 89055e0de0ff          mov [rip-0x1ff2a2], eax
000000010ea002a2: 4883c408              add rsp, 0x8
000000010ea002a6: c3                    ret
000000010ea002a7: 4983ee08              sub r14, 0x8
000000010ea002ab: 49c70601000000        mov qword [r14], 0x1
000000010ea002b2: 8905480de0ff          mov [rip-0x1ff2b8], eax
000000010ea002b8: 4883c408              add rsp, 0x8
000000010ea002bc: c3                    ret
000000010ea002bd: 0000                  add [rax], al
000000010ea002bf: 0000                  add [rax], al
000000010ea002c1: 0000                  add [rax], al
000000010ea002c3: 0000                  add [rax], al
000000010ea002c5: 0000                  add [rax], al
000000010ea002c7: 0000                  add [rax], al
000000010ea002c9: 0000                  add [rax], al
000000010ea002cb: 0000                  add [rax], al
000000010ea002cd: 0000                  add [rax], al
000000010ea002cf: 00                    invalid
```

New disassemble:

```
IN: scratchpad M\ fixnum bit? disassemble
000000010e656ec0: 89053ac1d6ff          mov [rip-0x293ec6], eax
000000010e656ec6: 4883ec08              sub rsp, 0x8
000000010e656eca: e8d10fe1ff            call 0x10e467ea0 ([ \ integer>fixnum-strict ~array~ 0 ~array~ inline-cache-miss ])
000000010e656ecf: 498b06                mov rax, [r14]
000000010e656ed2: 498b5ef8              mov rbx, [r14-0x8]
000000010e656ed6: 48c1fb04              sar rbx, 0x4
000000010e656eda: 48c1f804              sar rax, 0x4
000000010e656ede: 480fa3c3              bt rbx, rax
000000010e656ee2: 48bb0100000000000000  mov rbx, 0x1
000000010e656eec: 48b99c168f0101000000  mov rcx, 0x1018f169c
000000010e656ef6: 480f42d9              cmovb rbx, rcx
000000010e656efa: 4983ee08              sub r14, 0x8
000000010e656efe: 49891e                mov [r14], rbx
000000010e656f01: 8905f9c0d6ff          mov [rip-0x293f07], eax
000000010e656f07: 4883c408              add rsp, 0x8
000000010e656f0b: c3                    ret
000000010e656f0c: 0000                  add [rax], al
000000010e656f0e: 0000                  add [rax], al
```

**Comment by erg** [comment]
Looks fine. That's a pretty good speedup!

**Comment by mrjbq7** [comment]
:-)

If you call it directly, as part of other algorithms not as good a speedup.

**Comment by mrjbq7** [comment]
Maybe it would be a good blog post.

**Comment by mrjbq7** [comment]
Merged in 621b50a8e5c6beab78fd60c4c3a3d94f968406dd.
